### ISSUE TITLE
Error encountered during model fine-tuning

### ISSUE THREAD

**Issue by lpzoaa** [weberlab-hhu/Helixer#142]
Hi,
I am trying to fine-tuning the model with the command below.

`singularity run --nv ~/software/singularity/helixer.sif fasta2h5.py --species Ac --h5-output-path Ac.h5 --fasta-path Ac.fa`
`singularity run --nv ~/software/singularity/helixer.sif HybridModel.py --load-model-path $HOME/.local/share/Helixer/models/land_plant/land_plant_v0.3_a_0080.h5 --test-data Ac.h5 --overlap --val-test-batch-size 32 -v`
`singularity run --nv ~/software/singularity/helixer.sif helixer_post_bin Ac.h5 predictions.h5 100 0.1 0.8 60 Ac.post.gff3`
`singularity run --nv ~/software/singularity/helixer.sif import2geenuff.py --fasta Ac.fa --gff3 Ac.post.gff3 --db-path Ac.post.sqlite3 --log-file Ac.log --species Ac`
`singularity run --nv ~/software/singularity/helixer.sif geenuff2h5.py --h5-output-path Ac.post.h5 --input-db-path Ac.post.sqlite3 `
`python3 filter-to-most-certain.py --write-by 6415200 --h5-to-filter Ac.post.h5 --predictions predictions.h5 --keep-fraction 0.2 --output-file Ac.post.filtered.h5`
`mkdir -p Ac_data`
`python3 n90_train_val_split.py --write-by 6415200 --h5-to-split Ac.post.filtered.h5 --output-pfx Ac_data/`
`singularity run --nv ~/software/singularity/helixer.sif HybridModel.py -v --batch-size 50 --val-test-batch-size 100 -e100 --class-weights "[0.7, 1.6, 1.2, 1.2]" --transition-weights "[1, 12, 3, 1, 12, 3]" --learning-rate 0.0001 --resume-training --fine-tune --load-model-path $HOME/.local/share/Helixer/models/land_plant/land_plant_v0.3_a_0080.h5 --data-dir Ac_data --save-model-path ./Ac_best_model.h5`

But I encountered the following error while fine-tuning the model.

/usr/local/lib/python3.8/dist-packages/helixer/prediction/Metrics.py:74: RuntimeWarning: invalid value encountered in divide
  normalized_cm = self.cm / class_sums[:, None]  # expand by one dim so broadcast work properly

[slurm-9993101.log](https://github.com/user-attachments/files/17301661/slurm-9993101.log)
Could you please help me understand what might be causing this issue?
Thank you.

**Comment by felicitas215** [comment]
Hi,
thank you for posting this issue. I looked into your log file and the reference/Helixer annotation you use for fine tuning seems to be of poor quality. Your fine tuning run shows a perfect F1 score (1) for the intergenic regions and every other class (UTR, Exon, Intron) has a score of 0. These scores don't change during training. Looking at the confusion matrices, it seems that your data mostly contains integenic regions. Did you check the annotation gff3 file you got from Helixer or could you upload a zipped version of it?  Which species does "Ac" refer to?

**Comment by lpzoaa** [comment]
Hi,

"Ac" refers to onion (Allium cepa), which has a large genome size of 16 Gb. I have reviewed the GFF3 file obtained from Helixer and extracted gene sequences based on the provided annotations. Helixer performed well for our genome, successfully predicting nearly 80% of the genes, which is better compared to other de novo annotation tools.

I attempted to fine-tune the model using only one chromosome following the aforementioned process (since the entire genome is too large), but encountered an error. I suspect the issue lies in the fact that the reference, which is 2.3 Gb in size, contains only 5000 genes. This suggests that the intergenic regions make up the majority of the reference, and the 'filter-to-most-certain.py' script filtered out 20% of the results from Helixer, leaving only a small number of genes for training.

![image](https://github.com/user-attachments/assets/3732a853-7a01-421d-bd82-f08b56174747)

I also tried using gene annotations predicted by TransDecoder for training, and this worked successfully.

![image](https://github.com/user-attachments/assets/1b7eb576-1aaf-4119-9d47-512f70ee623d)

However, the new model I generated was only 8.5 MB in size, whereas the land_plant_v0.3_a_0080.h5 model is 25 MB. Additionally, the performance of the new model was not as good as the original—it only predicted 60% of the genes, which is even worse than before. Could this be because I only used one chromosome for training?

I would appreciate any advice you could offer.

**Comment by felicitas215** [comment]
Hi,
First of all, I'm happy to hear that Helixer performs well for your onion genome.
It could very well be that the fine tuning doesn't perform well, when you only use one chromosome instead of the entire genome. Is there a reason why you didn't use the entire genome? When retraining takes too long because of the genome size, you can try setting the `--batch-size` parameter higher than 50 (but just as high as your GPU allows, because it runs out of GPU RAM eventually, so maybe you need to try out different batch sizes).

**Comment by lpzoaa** [comment]
Hi,
I have retrained the model using the entire genome and tested the new model on part of the genome sequence. 

`singularity run --nv helixer.sif Helixer.py --lineage land_plant --fasta-path test.fa --model-filepath Ac_best_model.h5 --gff-output-path test.gff3 --subsequence-length 106920 --overlap-offset 53460 --overlap-core-length 80190`

Unfortunately, the performance is still suboptimal. Specifically, Helixer predicted about 500 genes with a BUSCO score of 2.2, but this genomic region contains approximately 1200 genes, and its BUSCO score is 3.4. Helixer's predictions seem to be very conservative, but we require higher sensitivity, as inaccurate gene predictions will be filtered out during the integration of multiple prediction results.

Here is the pipeline I used for training along with the log file:

here it's my pipeline and log file of training
`singularity run --nv helixer.sif HybridModel.py -v --batch-size 50 --val-test-batch-size 100 -e100 --class-weights "[0.7, 1.6, 1.2, 1.2]" --transition-weights "[1, 12, 3, 1, 12, 3]" --learning-rate 0.0001 --resume-training --fine-tune --predict-phase --load-model-path $HOME/.local/share/Helixer/models/land_plant/land_plant_v0.3_a_0080.h5 --data-dir Ac_data --save-model-path ./Ac_best_model.h5`
[slurm-73539123.out.gz](https://github.com/user-attachments/files/17373996/slurm-73539123.out.gz)
Could you please provide some advices on how I could improve the sensitivity of Helixer?
Thank you.

**Comment by alisandra** [comment]
Hi Ipzoaa, 

Some thoughts:
big picture 1) it sounds like no methods are performing to your satisfaction on your genome, do you have any idea why the genome is hard to annotate? Either biologically or technically, anything unusual/outlier-like about it? Some challenges might have specific fixes if one knows what they are. 

medium picture 2) Helixer (and any other DL model) is going to imitate the patterns in it's training data; so the better data you can take for fine-tuning (in regards to your use case), the better. For instance, it sounds like you might want to consider greatly increasing the `--keep-fraction` when running `filter-to-most-certain.py` (e.g. to 0.5 or 0.8, or just skip the step entirely). Similarly, if they're anyways good, you can consider simply taking labels from TransDecoder / any other source. 

focus 3) assuming you're retraining/fine tuning, you have a lot of control over precision/recall balance via the class weights parameter. You might try things that give intergenic a much lower weight, such as `--class-weights "[0.4, 1.6, 1.2, 1.2]"` or even `--class-weights "[0.1, 1.6, 1.2, 1.2]"`. This should increase the recall at the cost of the precision. 

focus 4) during _inference_ you can also try a lower "peak-threshold" argument (to HelixerPost), e.g. 0.7 or 0.75

misc tip 5) when re-training, you can add the parameter `--check-every-nth-batch` and set it to something smaller than your batches per epoch (e.g. 2000), such that it checks for convergence more often & there's less overhead in time required for training with more data.

**Comment by lpzoaa** [comment]
Hi Alisandra,

Thank you for your reply. I believe that the methods generally do not perform well for large genomes (>10 Gb), not just for mine, since transposable elements (TEs) and intergenic regions make up the majority of such genomes.

I have tried assigning a lower weight to the intergenic regions using --class-weights "[0.1, 1.6, 1.2, 1.2]" and also set the --peak-threshold to 0.7. With these adjustments, I obtained more gene annotations compared to the default settings. However, the BUSCO score did not improve.

I think de novo annotation of large genomes is still a challenging task. Helixer has been one of the best-performing tools among those we have tried.  I hope you could achieve breakthroughs in this area in your future research.

best wishes
pengzheng

**Comment by alisandra** [comment]
Thanks for sharing your learnings with us and the encouragement. I hope you find something that fits your needs soon!
